tests/test_common.py

```diff
@@ -326,6 +338,38 @@ def test_join(df_raw: Any) -> None:
     compare_dicts(result_native, expected)
 
 
+# Sample data
+df_pandas = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+df_right_pandas = pd.DataFrame({"a": [1, 2, 3], "c": [4, 5, 6]})
+
+df_polars = pl.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+df_right_polars = pl.DataFrame({"a": [1, 2, 3], "c": [4, 5, 6]})
+
+
+@pytest.mark.parametrize(
+    "df_left, df_right", [(df_pandas, df_right_pandas), (df_polars, df_right_polars)]
+)
+def test_coalesce_overlapping_names(df_left, df_right):
+    if isinstance(df_left, pd.DataFrame):
+        # Perform the join operation in pandas
+        result_pandas = df_left.merge(df_right, left_on="b", right_on="c", how="left")
+        # Adjust the pandas result to match the expected column names
+        result_pandas = result_pandas.rename(columns={"a_x": "a_left", "a_y": "a_right"})
+        result = result_pandas.to_dict(orient="list")
+    else:
+        # Perform the join operation in narwhals (using polars underneath)
+        df_left_nw = nw.DataFrame(df_left)
+        df_right_nw = nw.DataFrame(df_right)
+        result_polars = df_left_nw.join(
+            df_right_nw, left_on="b", right_on="c", how="left"
+        )
+        result = nw.to_native(result_polars)
+        result = result.rename({"a": "a_left"}).to_pandas().to_dict(orient="list")
+
+    expected = {"a_left": [1, 2, 3], "b": [4, 5, 6], "a_right": [1, 2, 3], "c": [4, 5, 6]}
+    compare_dicts(result, expected)
```

Comment:
the structure of the test should be more like:

- given
- when
- then

So:
- given: `df_pandas`, `df_pandas_right`, `df_polars`, `df_polars_right` (you've done this right)
- then: you do
  ```python
  df_left = nw.from_native(df_left)
  df_right = nw.from_native(df_right)
  result = df_left.join(df_right, left_on='b', right_on='c', how='left')
  ```
- then: you make `expected`, and check that `result` matches it

There shouldn't be any pandas-specific logic here. If there is, then it needs adjusting in  https://github.com/narwhals-dev/narwhals/blob/a22b600595559d9ba7950e3bc76badd4be76a548/narwhals/_pandas_like/dataframe.py#L249-L270